### PR TITLE
New Plugin API PR3: Introduces Service, BroadcastReceiver, and ContentProvider awareness, control surfaces, and plugin bindings.

### PR DESCRIPTION
**Pull request by matthew-carroll**
New Plugin API PR3: Introduces Service, BroadcastReceiver, and ContentProvider awareness, control surfaces, and plugin bindings.

In this PR I took the existing FlutterEnginePluginRegistry, which was defined within the FlutterEngine file, and broke it out to a package private standalone class. Most if its content is unchanged.

This PR repeats a structural process for `Services`s, `BroadcastReceiver`s, and `ContentProvider`s that was done for `Activity`s in the previous PR:
https://github.com/flutter/engine/pull/8943

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.plugins.broadcastreceiver;
+
+import android.support.annotation.NonNull;
+
+/**
+ * {@link FlutterPlugin} that wants to know when it is running within a {@link BroadcastReceiver}.
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java]: nit: Having trouble parsing this phrase.

Maybe:

```suggestion
 * A {@link FlutterPlugin} that wants to know when it is running within a {@link BroadcastReceiver}.
```

File: shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ */
+public interface BroadcastReceiverAware {
+  /**
+   * This {@code BroadcastReceiverAware} {@link FlutterPlugin} is now associated with a
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java]: nit: passive voice, also see [the summary fragment](https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment). Applies  to the below method too. 

```suggestion
   * Callback triggered when a {@code BroadcastReceiverAware} {@link FlutterPlugin} is associated with a {@link BroadcastReceiver}.
```

File: shell/platform/android/io/flutter/embedding/engine/plugins/service/ServiceAware.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.plugins.service;
+
+import android.support.annotation.NonNull;
+
+/**
+ * {@link FlutterPlugin} that wants to know when it is running within a {@link Service}.
+ */
+public interface ServiceAware {
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/engine/plugins/service/ServiceAware.java]: Same comments here as on `BroadcastReceiverAware`.

File: shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java (hunk 1 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java]: done

File: shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java (hunk 2 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/plugins/broadcastreceiver/BroadcastReceiverAware.java]: done

File: shell/platform/android/io/flutter/embedding/engine/plugins/service/ServiceAware.java (hunk 3 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/plugins/service/ServiceAware.java]: done

**Pull request merged**